### ISSUE TITLE
Change Request: Reject non string replacements in `fixer.replaceText()`

### ISSUE THREAD

**Issue by fisker** [eslint/eslint#18807]
### ESLint version

v9

### What problem do you want to solve?

In a rule `fix` function, 

```js
fixer.replaceText(node, replacement.text);
```

when `replacement.text` is undefined, it will replace the node with `undefined`.

### What do you think is the correct solution?

Throw an error instead.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

Prevent mistakes.

**Comment by snitin315** [comment]
Throwing an error here would be a breaking change.

**Comment by nzakas** [comment]
I think this is a good idea. Marking as breaking for now, though I think there's an argument to be made that this is fixing a bug so perhaps we could do it in a minor release? @eslint/eslint-team 

@snitin315 please make sure you're moving issues through the categories on the Triage board. Comments are helpful, but we want to make sure each issue is progressing through the triage process.

**Comment by JoshuaKGoldberg** [comment]
👍, there's no reason to replace with anything that's not `typeof` `"string"`. Using it that way would be a bug IMO.

How comprehensive do we want to be in this change? There are a lot of public fixer APIs that take strings - all but `remove` & `removeRange` in:

https://github.com/eslint/eslint/blob/de408743b5c3fc25ebd7ef5fb11ab49ab4d06c36/docs/src/extend/custom-rules.md#L292-L299

...not to mention other APIs outside of fixers.

Proposal: apply for this specific issue, have it apply to the `text` in all public fixer methods?

**Comment by mdjermanovic** [comment]
I also think we should treat this as a breaking change. `fixer.replaceText(node, undefined);` currently works by replacing the node text with `"undefined"`, and although unlikely, such code might be intentional (or at least producing the intended result) in some cases.

**Comment by nzakas** [comment]
All right, adding to the v10.0.0 board for tracking.

**Comment by mdjermanovic** [comment]
> How comprehensive do we want to be in this change? There are a lot of public fixer APIs that take strings

I think it should apply to all fixer methods that take strings.

**Comment by Jay-Karia** [comment]
What should be the error "type" and "message" ?

**Comment by fisker** [comment]
```js
throw new TypeError("Expected 'parameter-name' to be a string, got 'undefined'.")
```

**Comment by Jay-Karia** [comment]
As far as I can understand, this is the place where we need to check the value of `text`.

https://github.com/eslint/eslint/blob/e849dc01286cde5b6e2f0e04bf36928710633715/lib/linter/rule-fixer.js#L113-L117

Am I on the right path ?

**Comment by nzakas** [comment]
@Jay-Karia yes. Keep in mind that this feature is planned for v10 because it's a breaking change, and we don't yet have a schedule for v10, which is why work hasn't started on it yet.

**Comment by sam3k** [comment]
TSC group has confirmed this for `v10`

**Comment by Pixel998** [comment]
I'll work on this.
